### PR TITLE
Tests for proposal pruning across epochs

### LINKED ISSUE #3894: Test unreachable proposals clean up functionality
Following #3855 write sanity unit tests to cover the functionality.

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Resolves #3894 

This is split into two and superseded by #3978 and #3981 

# Pending
- [x] Isolate all operations over governance proposals in `Proposals` module
- [x] Write property tests for all functionality concerning `Proposals` addition, expiry and enactment
- [ ] Write unit tests to cover all scenarios
  - [x] Basic
  - [ ] Advanced
- [x] Remove all traces and debugging code from this PR
- [x] Clean every file that has been touched
- [x] Fix roundtrip tests
- [x] Update `cardano-ledger-test`
- [ ] Document the `Conway.Governance.Proposals` module
- [x] Rearrange commit sequence

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 1]
```diff
@@ -609,6 +613,22 @@ passEpoch = do
...
+  , STS (EraRule "TICK" era)
+  , Signal (EraRule "TICK" era) ~ SlotNo
+  , NFData (State (EraRule "TICK" era))
+  , State (EraRule "TICK" era) ~ NewEpochState era
+  ) =>
+  Word8 ->
+  ImpTestM era ()
+passNEpochs = \case
+  0 -> pure ()
+  n -> passEpoch >> passNEpochs (n - 1)
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Let me ask you, what will happen if I accidentally pass `-1` to this function? There is a reason why we still use `Int` for sizes and indices, cause unsigned will happily overflow for negative numbers leading to bad or wird bugs

```suggestion
  Natural ->
  ImpTestM era ()
passNEpochs n = when (n > 0) $ passEpoch >> passNEpochs (n - 1)
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 2]
```diff
@@ -389,6 +392,7 @@ newtype ImpTestM era a = ImpTestM (ReaderT (ImpTestEnv era) IO a)
     , MonadIO
     , MonadUnliftIO
     , MonadReader (ImpTestEnv era)
+    , MonadFail
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: It is better to provide a manual instance, because `IOException` is not very informative
```suggestion
```

```haskell
instance MonadFail (ImpTestM era) where
  fail = expectationFailure
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs [hunk 3]
```diff
@@ -1,5 +1,4 @@
 module Cardano.Ledger.Conway.Core (
-  ConwayEraTxBody (..),
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs:2]: Please, bring back all the changes you made to this module

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs [hunk 4]
```diff
@@ -502,8 +461,12 @@ instance EraPParams era => NoThunks (EnactState era)
...
 data RatifyState era = RatifyState
   { rsEnactState :: !(EnactState era)
-  , rsRemoved :: !(Set (GovActionId (EraCrypto era)))
-  , rsEnacted :: !(Set (GovActionId (EraCrypto era)))
+  , rsEnacted :: !(Seq (GovActionId (EraCrypto era)))
+  , rsExpired :: !(Set (GovActionId (EraCrypto era)))
+  -- ^ Sequence of actions that have been enacted at an epoch
+  -- boundary. This is used temporarily as a sequence to be replayed
+  -- on Proposals. This is again reset at the epoch boundary, once we
+  -- have replayed the enactments in order onto the Proposals
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: The whole `RatifyState` is temporary. We only produce it as a result of DRep pulser. So this comment is not very useful
```suggestion
  -- boundary.
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs [hunk 5]
```diff
@@ -1106,6 +1080,8 @@ setFreshDRepPulsingState ::
   ( GovState era ~ ConwayGovState era
   , Monad m
   , RunConwayRatify era
+  , Era era
+  -- , ConwayEraGov era
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs [hunk 6]
```diff
@@ -139,7 +139,11 @@ data GovActionId c = GovActionId
   { gaidTxId :: !(TxId c)
   , gaidGovActionIx :: !GovActionIx
   }
-  deriving (Generic, Eq, Ord, Show)
+  deriving (Generic, Eq, Ord)
+
+instance Show (GovActionId c) where
+  show (GovActionId txid (GovActionIx gaix)) =
+    take 6 (drop 1 $ show $ extractHash $ unTxId txid) <> take 4 (show gaix)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs]: Please, use derived instances for Show. We want to see the whole contents of the type
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs [hunk 7]
```diff
@@ -189,9 +193,18 @@ data GovActionState era = GovActionState
...
+
+instance EraPParams era => Show (GovActionState era) where
+  show (GovActionState {gasId, gasAction, gasExpiresAfter}) =
+    unlines
+      [ "----- GovActionState -----"
+      , show gasId
+      , show gasAction
+      , show gasExpiresAfter
+      , "----- GovActionState end -----"
+      ]
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs]: Derived instances. Custom formatting is reserved for other type classes, not for `Show`
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs [hunk 8]
```diff
@@ -546,36 +551,44 @@ data GovActionPurpose
...
-    , NFData
-    , ToJSON
-    , Ord
+    , -- , Show
+      -- , EncCBOR
+      -- , DecCBOR
+      -- , NoThunks
+      -- , NFData
+      -- , ToJSON
+      Ord
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs]:
```suggestion
    , Ord
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs [hunk 9]
```diff
@@ -598,14 +611,29 @@ data GovAction era
...
+    TreasuryWithdrawals _ ->
+      "TreasuryWithdrawals (-)"
+    NoConfidence parent ->
+      "NoConfidence (" <> show parent <> ")"
+    UpdateCommittee parent _ _ _ ->
+      "UpdateCommittee (" <> show parent <> ")"
+    NewConstitution parent _ ->
+      "NewConstitution (" <> show parent <> ")"
+    InfoAction ->
+      "InfoAction (-)"
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs]: Ditto. Derive Show

```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 10]
```diff
@@ -1,132 +1,452 @@
...
+          ]
+      , mconcat
+          [ "<C"
+          , "{"
+          , show $ Set.size children
+          , "}"
+          , show $ Set.toList children
+          , ">"
+          ]
+      ]
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 11]
```diff
@@ -1,132 +1,452 @@
...
+      , case parent of
+          Just p -> mconcat ["(", show p, ")"]
+          Nothing -> "(-)"
+      , ","
+      , "{"
+      , show $ Set.size children
+      , "}"
+      , show $ Set.toList children
+      , ">"
+      ]
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 12]
```diff
@@ -1,132 +1,452 @@
...
+
+_pnParentL :: Lens' (PNode a) (Maybe a)
+_pnParentL = lens pnParent $ \x y -> x {pnParent = y}
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxow.hs [hunk 13]
```diff
@@ -41,6 +41,7 @@ import Cardano.Ledger.BaseTypes (ShelleyBase)
 import Cardano.Ledger.Conway.Core
 import Cardano.Ledger.Conway.Era (ConwayUTXOW)
 import Cardano.Ledger.Conway.Governance (Voter (..), VotingProcedures (..))
+import Cardano.Ledger.Conway.TxBody (ConwayEraTxBody (votingProceduresTxBodyL))
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxow.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxos.hs [hunk 14]
```diff
@@ -35,6 +35,7 @@ import Cardano.Ledger.BaseTypes (ShelleyBase)
 import Cardano.Ledger.Conway.Core
 import Cardano.Ledger.Conway.Era (ConwayUTXOS)
 import Cardano.Ledger.Conway.Governance (ConwayGovState (..))
+import Cardano.Ledger.Conway.TxBody (ConwayEraTxBody (treasuryDonationTxBodyL))
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxos.hs:38]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 15]
```diff
@@ -1,132 +1,452 @@
...
+pnChildrenL = lens pnChildren $ \x y -> x {pnChildren = y}
+
+newtype PHierarchy a = PHierarchy
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+instance Show a => Show (PHierarchy a) where
+  show = show . Map.assocs . unPHierarchy
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]: Hiding the newtype wrapper in Show instance is wrong
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 16]
```diff
@@ -1,132 +1,452 @@
...
+      , show pfPParamUpdate
+      , ">pfHardFork"
+      , show pfHardFork
+      , ">pfCommittee"
+      , show pfCommittee
+      , ">pfConstitution"
+      , show pfConstitution
+      , "----- PForest end -----"
+      ]
+
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]:
```suggestion
```

**Review by lehins — COMMENTED** [review]
Sorry, I didn't get a chance to look through the whole PR yet, but here is some feedback that I can already provide

File: eras/conway/impl/cardano-ledger-conway.cabal [hunk 17]
```diff
@@ -96,16 +100,20 @@ library
...
+        ghc-options: -fno-ignore-asserts
+
 library testlib
     exposed-modules:
         Test.Cardano.Ledger.Conway.Arbitrary
         Test.Cardano.Ledger.Conway.Binary.Cddl
         Test.Cardano.Ledger.Conway.Binary.RoundTrip
         Test.Cardano.Ledger.Conway.ImpTest
         Test.Cardano.Ledger.Conway.Imp
+        Test.Cardano.Ledger.Conway.Imp.GovSpec
```
Comment (teodanciu) [inline eras/conway/impl/cardano-ledger-conway.cabal]: I wonder why you moved this here, it seems to make more alphabetical sense after `EnactState` where it was, no?

Comment (aniketd) [reply eras/conway/impl/cardano-ledger-conway.cabal]: Yes agree :100: . The previous position has been restored. :raised_hands:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs (hunk 4 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Removed! :100:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs (hunk 6 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs]: :100: I will revert all the custom `Show` instances. It was helpful for tracing, but its now time to get rid of them :+1:

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/GovActionReorderSpec.hs [hunk 18]
```diff
@@ -2,40 +2,42 @@
...
-      actionsList <-
-        traverse
-          (>>= genGovActionStateFromAction)
-          (govActionGenerators @Conway)
-      let sortedActions = sortOn (actionPriority . gasAction) actionsList
-      shuffledActions <- shuffle actionsList
-      pure $
-        Seq.fromList sortedActions
-          == reorderActions (Seq.fromList shuffledActions)
+spec = undefined
```
Comment (Soupstraw) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/GovActionReorderSpec.hs]: Why is this here?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 19]
```diff
@@ -1,132 +1,451 @@
...
+deriving instance Era era => Default (PForest PHierarchy era)
+
+data Proposals era = Proposals
+  { pOMap :: !(OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
+  , pRoots :: !(PForest PRoot era)
+  , pHierarchy :: !(PForest PHierarchy era)
+  }
+  deriving (Eq, Generic, NoThunks, NFData, Default)
+
+instance Era era => Show (Proposals era) where
```
Comment (aniketd) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]: This custom instance has also been removed.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 20]
```diff
@@ -140,264 +459,307 @@ proposalsAddVote voter vote gai (Proposals omap) =
...
+
+deriving instance Era era => Eq (PForest StrictMaybe era)
+deriving instance Era era => Ord (PForest StrictMaybe era)
+deriving instance Era era => Generic (PForest StrictMaybe era)
+deriving instance Era era => NoThunks (PForest StrictMaybe era)
+deriving instance Era era => NFData (PForest StrictMaybe era)
+deriving instance Era era => ToJSON (PForest StrictMaybe era)
+deriving instance Era era => Default (PForest StrictMaybe era)
+
+newtype PrevGovActionIds era = PrevGovActionIds
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:677]: I wonder if we shouldn't rename this as well, since we eradicated `PrevGovActionId` . Maybe `GovPurposeIds` ?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 21]
```diff
@@ -1,132 +1,451 @@
...
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+instance Show a => Show (PHierarchy a) where
+  show = show . Map.assocs . unPHierarchy
+
+pHierarchyNTL :: Lens' (PHierarchy a) (Map a (PNode a)) -- FIXME: what a terrible name!
+pHierarchyNTL = lens unPHierarchy $ \x y -> x {unPHierarchy = y}
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: What is `NT`  ?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 22]
```diff
@@ -1,132 +1,451 @@
...
-  def = Proposals def
+deriving instance Era era => Eq (PForest PHierarchy era)
+deriving instance Era era => Ord (PForest PHierarchy era)
+deriving instance Era era => Generic (PForest PHierarchy era)
+deriving instance Era era => NoThunks (PForest PHierarchy era)
+deriving instance Era era => NFData (PForest PHierarchy era)
+deriving instance Era era => Default (PForest PHierarchy era)
+
+data Proposals era = Proposals
+  { pOMap :: !(OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]: Not sure about this name, it expresses what's not so necessary (the type) and nothing about its meaning.  Should we go back to the `govActionsState` that we had at some point before?

Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]: 100%
Naming fields with types is an antipattern

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 23]
```diff
@@ -1,132 +1,451 @@
...
+            SNothing ->
+              -- We only accept actions with parent `SNothing` when
+              -- there are no proposals enacted yet, which means that
+              -- the root has to be `Nothing`
+              if isNothing $ ps ^. pRootsL . lenzR . prRootL
+                then
+                  Just $
+                    ps
+                      & pOMapL %~ (OMap.||> gas)
+                      & pRootsL . lenzR . prChildrenL %~ Set.insert newId
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:270]: What do you think about having a few variables with meaningful names for some of these expressions that are reused often?  
Like: 
`pRootsL . lenzR . prChildrenL`
`pHierarchyL . lenzH . pHierarchyNTL` 

Maybe it will make it easier to understand the essence of this algorithm?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs [hunk 24]
```diff
@@ -1,132 +1,451 @@
...
+    NewConstitution parent _ ->
+      update pfrConstitutionL pfhConstitutionL parent
+    InfoAction -> Just $ ps & pOMapL %~ (OMap.||> gas)
+  where
+    update ::
+      Lens' (PForest PRoot era) (PRoot (GovPurposeId p era)) ->
+      Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+      StrictMaybe (GovPurposeId p era) ->
+      Maybe (Proposals era)
+    update lenzR lenzH prev =
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:257]: Out of curiosity, why did you opt for `z` in `lenz` instead of `s` ? Is it to avoid a name collision?

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 1 above)
Comment (aniketd) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: :100: Thank you!

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 2 above)
Comment (aniketd) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: `assertFailure` did the trick. Thank you!

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/GovActionReorderSpec.hs (hunk 18 above)
Comment (aniketd) [reply eras/conway/impl/test/Test/Cardano/Ledger/Conway/GovActionReorderSpec.hs]: These tests have now been restored. :+1:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 20 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:677]: The `Prev` prefix was originally intended to mean that these were enacted on the previous epoch boundary. Since there is no semantic change to this type, keeping it as is, one could argue, still does that job of conveying this meaning.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 21 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: :vomiting_face: NewType

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: I want a recommendation to name this differently :yum: Can you think of a name, especially considerting that `pHierarchyL` is already taken. cc. @Soupstraw

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 20 above)
Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:677]: Oh I see, that makes sense.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 22 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs]: The new name now is `pProps`. Unless you guys think `pActionStates` is a better alternative.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 21 above)
Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: Maybe `nodesPHierarchyL` ?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 23 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:270]: Here `lenzR` and `lenzH` are polymorphic, and also passed into the scope from the calling site. Do you mean having a `let` binding like the following would help? I'd be happy to do this if you confirm that I understood your suggestion correctly. :grin: 
```haskell
let childrenOfRootL = pRootsL . lenzR . prChildrenL
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 24 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:257]: Yes! It is to avoid name collision.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs (hunk 21 above)
Comment (Soupstraw) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: hmm, maybe rename `pHierarchy` to `pHierarchyForest` and then rename this to `pHierarchy`?
Or use `pHierarchyNodes` here.

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs:149]: Nice thanks! :raised_hands: I will use this :+1:

**Comment by lehins** [comment]
Closing in favor of #3981
